### ISSUE TITLE
Enable change push / sync to Atlas and possibly trigger builds (CI)

### ISSUE THREAD

**Issue by lenucksi** [InnerSourceCommons/InnerSourceLearningPath#108]
# Rationale
O'Reilly Production happens in Atlas. Atlas conceptually is a large GitLab instance with attached document production capabilities. We want to enable push of our upstream changes to the O'Reilly system to enable faster and easier collaboration for everyone while ensuring they do not break anything at O'Reilly.
Thus we need to sync our repo on changes and check PR's to not break the O'Reilly process - in other words: We need CI.

# Action points
- Figure out Atlas build process
  - build configuration files, jobs (separation of articles)
  - API usage to trigger artifact builds, possibly provide artifacts?
- Figure out possibly needed changes to format to enable good looking results for PDF and EPUB.
- Figure out credential storage, appropriate CI system, possibly acquire needed non-personal accounts
- Setup jobs on testing copy and validate with O'Reilly that nothing breaks
  - PR tests
  - Continuous sync/push of ```master``` to O'Reilly
- Write about the approach s.t. it can possibly be useful for others.

# Possible additional use
- Trigger builds to GitBook or similar for our own use / site rendering
  - possibly also relevant for video integration (#46, #66)
- Possibly partially also useful for the patterns repo.

**Comment by rrrutledge** [comment]
I'm not sure about spending the time to do all this when we just have a few articles left to write.

**Comment by lenucksi** [comment]
> I'm not sure about spending the time to do all this when we just have a few articles left to write.

I'll spend a few cycles there to see how much effort it is and how re-useable for our own re-use that ends up being. 
Might even boil down to a quick git push GitHub -> Atlas on change of master action or something similar. And if it's GitLab anyway maybe we can even see if GitLab's mirror functionality is available.

**Comment by lenucksi** [comment]
## Rationale
A quick look through the available options revealed that GitHub Actions (currently in beta, should work for our small use case) might be very nice since it allows us to react to events such as a commit and even provides regular time-based scheduled execution triggers for free in our size.

## Branches
I see that two repos exist with a few branches each that we might want to connect and sync by pushing and pulling regularly.

### O'Reilly repo
- upstream -> InnerSource Commons master
- master -> Their production master
- editors -> Changes proposed by O'Reilly's editors

### InnerSource Commons repo
- master -> InnerSource Commons master
- oreilly-master -> O'Reilly's master branch
- oreilly-editors -> O'Reilly's editors branch

## Actions
### On commit to GitHub master (Individual commit/push or merge of a PR)
- push to upstream at O'Reilly
- build PDF from upstream-branch via the Atlas-System for review and validation e.g. for #114 reasons.
- Future use: Build+deploy GitHub-site or whatever system and format we choose for our rendering of the material
- Possibly announce the results on Slack.

### On commit to O'Reilly master
- Technically Atlas could realize this as it's a GitLab instance. I don't think we'd get access to these features though.
  - push both O'Reilly branches to GitHub
  - Build and announce review copies
- Alternatively we could use cron'd triggers, e.g. on a 15 minute basis to poll the O'Reilly 

## Required resources
- We'd need to create&deploy a SSH-deploy-key secret for the O'Reilly repo. As far as I've understood the GitHub actions documentation, owners of the GitHub repo or the GitHub org would have access to that secret. 
- A separate bot account at O'Reilly might be nice.

**Comment by rrrutledge** [comment]
Why not call `upstream` something like `learningpath-master` instead to have consistent way of naming remote branches?

**Comment by lenucksi** [comment]
> Why not call `upstream` something like `learningpath-master` instead to have consistent way of naming remote branches?

```upstream``` would be the canonical OSS term, thus I used it. I agree with your argument though, changed it.

**Comment by rrrutledge** [comment]
LGTM

**Comment by lenucksi** [comment]
Relevant raw scripting is here: https://github.com/lenucksi/InnerSourceLearningPath-AtlasIntegration/tree/master/.github/workflows

This needs to be done
- Find the relevant O'Reilly repos
 - Looks like each chapter goes into a separate repo, thus syncing either branches or the entire repo to multiple separate repos might be required.
- Find the relevant corresponding branches
 - Very likely separate naming will not be required but automatic namespacing such as oreilly/master, oreilly/editors in our repo and github/master in theirs should suffice.
- Ensure presence of a master and review/editors branch on their side.
- Ensure the relevant build config is present there (atlas.json + templates)
  - Existing config can likely be re-used or provided by O'Reilly
- As far as it looks there's an action to automatically create PRs from updates to the oreilly/master or oreilly/editors branches in our repo.
- Steps to build our rendering from the existing material need to be added as a separate action and triggered on push/PR to master.

Possible workflow after that would be 
- sync of new edits from Atlas here
- reviewing of PRs from oreilly/editors -> master by everyone here
- sync of our master back to Atlas as github/master
- merge to their master / editors branch to be incorporated.

Current actions in the repo given above fail in one direction since the repos went out of sync due to manual edits.

If anyone is interested in participating in doing/automating the remaining steps I'd be happy.

**Comment by rrrutledge** [comment]
I think as a first step if we could write the syncing scripts even if they have to be executed manually that would provide some value.  Anytime we want to sync things up someone with access to both repos can run the scripts.

**Comment by lenucksi** [comment]
> I think as a first step if we could write the syncing scripts even if they have to be executed manually that would provide some value.

Then just take what the scripts mentioned above do and do that manually. 

Since there's a human being in front of the box in this case, less error handling will be required and sync is more flexible - or error prone - however one wants to interpret the automated vs. human-handled difference. ;)

**Comment by rrrutledge** [comment]
This PR has some work in this area: https://github.com/InnerSourceCommons/InnerSourceLearningPath/pull/190

**Comment by lenucksi** [comment]
The pipeline to sync master to Atlas and build the current master on Atlas builds clean for the introduction section as of https://github.com/InnerSourceCommons/InnerSourceLearningPath/commit/ad721f6b08a2af8ee73532101433f7d5b6a0e36d.

Remaining work:
- Copy, paste and adapt this workflow to the other sections and the workbook. Adding [path filters](  https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#onpushpull_requestpaths) is likely beneficial to save on build minute and Atlas usage.
- To enable building and validating PDFs for pull requests we need push the respective pull request branch from GitHub into Atlas and have Atlas build off of that branch. Should be doable in a few hours of work.

**Comment by lenucksi** [comment]
Approach discussed to be used is as follows:
- The O'Reilly editors will conduct the edit / reviews using our repo and the GitHub web editors. Doable once #187 is done.
- Review results will be technically validated using CI using a Atlas as test system, as illustrated here. If Atlas is not used, [epubcheck](https://github.com/w3c/epubcheck) would be the required linting tool.
- Once reviews merge into master, everything is synced to Atlas repos and built there. See our [CI here](https://github.com/InnerSourceCommons/InnerSourceLearningPath/blob/master/.github/workflows/SyncToAtlas.yml), too.
- Every repo at Atlas is a clone of our repo, only differing in build related content such as templates, see [here](https://github.com/InnerSourceCommons/InnerSourceLearningPath/pull/190/files#r346903632). We will use one repo for each section plus one for the workbook.
- Target artifacts are PDF and EPUB.
- The editors at O'Reilly might have questions on how to use our repo, we can to help them with that.
- Nick at O'Reilly is our technical contact regarding Atlas.
